core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala

```diff
@@ -486,23 +492,33 @@ class HttpGoogleServicesDAO(
     }
   }
 
-  override def copyFile(sourceBucket: String, sourceObject: String, destinationBucket: String, destinationObject: String): Future[Option[StorageObject]] = {
+  override def copyFile(sourceBucket: String, sourceObject: String, destinationBucket: String, destinationObject: String, userProject: Option[GoogleProjectId]): Future[Option[StorageObject]] = {
     implicit val service = GoogleInstrumentedService.Storage
-
     val copier = getStorage(getBucketServiceAccountCredential).objects.copy(sourceBucket, sourceObject, destinationBucket, destinationObject, new StorageObject())
-    retryWithRecoverWhen500orGoogleError(() => { Option(executeGoogleRequest(copier)) }) {
+
+    val copierWithProject = userProject match {
+      case Some(project) => copier.setUserProject(project.value)
+      case None => copier
+    }
+
+    retryWithRecoverWhen500orGoogleError(() => { Option(executeGoogleRequest(copierWithProject)) }) {
       case e: HttpResponseException => {
         logger.warn(s"encountered error [${e.getStatusMessage}] with status code [${e.getStatusCode}] when copying [$sourceBucket/$sourceObject] to [$destinationBucket]")
         None
       }
     }
   }
 
-  override def listObjectsWithPrefix(bucketName: String, objectNamePrefix: String): Future[List[StorageObject]] = {
+  override def listObjectsWithPrefix(bucketName: String, objectNamePrefix: String, userProject: Option[GoogleProjectId]): Future[List[StorageObject]] = {
     implicit val service = GoogleInstrumentedService.Storage
     val getter = getStorage(getBucketServiceAccountCredential).objects().list(bucketName).setPrefix(objectNamePrefix).setMaxResults(maxPageSize.toLong)
 
-    listObjectsRecursive(getter) map { pagesOption =>
+    val getterWithProject = userProject match {
+      case Some(project) => getter.setUserProject(project.value)
```

Comment:
One thing to be careful of is that this means the user project will pay even when the bucket is not RP. I am not saying it is wrong as written just that it is important to know how this works.

Reply:
This is a good point - I will talk to our PO to make sure this is understood and acceptable!